// hikki-me/jamal-bot/src/Types/scraper.ts
export interface ITiktok {
  title: string
  keywords: string[]
  description: string
  canonicalHref: string
  robotsContent: string
  applicableDevice: string
  id: string
  desc: string
  createTime: number
  scheduleTime: number
  video: {
    id: string
    height: number
    width: number
    duration: number
    ratio: number
    cover: string
    originCover: string
    dynamicCover: string
    playAddr: string
    downloadAddr: string
    shareCover: string[]
    reflowCover: string
    bitrate: number
    encodedType: string
    format: string
    videoQuality: string
    encodeUserTag: string
    codecType: string
    definition: string
  }
  author: {
    id: string
    shortId: string
    uniqueId: string
    nickname: string
    avatarLarger: string
    avatarMedium: string
    avatarThumb: string
    signature: string
    verified: boolean
    secUid: string
  }
  music: {
    id: string
    title: string
    playUrl: string
    coverLarge: string
    coverMedium: string
    coverThumb: string
    authorName: string
    original: boolean
    duration: number
    album: string
    scheduleSearchTime: string
  }
  stats: {
    diggCount: number
    shareCount: number
    commentCount: number
    playCount: number
  }
  secret: boolean
  forFriend: boolean
  digged: boolean
  itemCommentStatus: number
  showNotPass: boolean
  vl1: boolean
  authorStats: {
    followerCount: number
    followingCount: number
    heart: number
    heartCount: number
    videoCount: number
    diggCount: number
  }
  diversificationLabels: string[]
  indexEnabled: boolean
}
