// hikki-me/jamal-bot/src/lib/api/meta-scraper.ts
import cheerio from 'cheerio'
import axios from 'axios'
import fetch from 'node-fetch'
import fromdata from 'form-data'
import { writeFileSync } from 'fs'
export class metaScraper {
  async getToken(url: string): Promise<string> {
    return new Promise(async (resolve, reject) => {
      const html = await axios.get('https://downvideo.quora-wiki.com/tiktok-video-downloader#url=' + url)
      const $ = cheerio.load(html.data)
      const token = $('#token').attr('value')
      resolve(token)
    })
  }
  public async scraperMeta(url: string) {
    return new Promise(async (resolve, reject) => {
      let form = new fromdata()
      const token = await this.getToken(url)
      form.append('url', url)
      form.append('token', token)
      const data = await fetch('https://downvideo.quora-wiki.com/system/action.php', {
        method: 'POST',
        body: form,
        headers: {
          Cookie: 'fpestid=N-NFBb48WFH1mPQZecUc2Q9_jaYVEBXTIUGihxu3Y-IMt2x73fD8rfJVGtgf_EocY6SSiQ; __gads=ID=30ab1aeee76ddf81-22a80ed95cd20056:T=1650467795:RT=1650467795:S=ALNI_MaRwGYVfTZzcJ_NqLu_can9yH-OtQ; PHPSESSID=akcq346g6a0ptkv8nlhe64f1ji',
          'Content-Type': 'multipart/form-data; boundary=' + form.getBoundary(),
          'Content-Length': form.getLengthSync().toString(),
        },
      })
      if (data.status !== 200) return reject(data.status)
      const json = await data.json()
      resolve(json)
    })
  }
  public async grphqlApi(url: string): Promise<any> {
    return new Promise(async (resolve, reject) => {
      axios({
        url: url.split(/[?]/)[0] + '?__a=1',
        method: 'GET',
        headers: {
          cookie: 'ig_did=F9A902C2-32C3-44DE-8697-C6420399B1AF; ig_nrcb=1; mid=Yi3zgwALAAGX5uTa-3DzYhbNz8iZ; csrftoken=1gIWQvgaJzSBeeyetkTnDccp9FPTmxvI; ds_user_id=41397968341; sessionid=41397968341:m4b9a9NtMEXXH1:2;',
          'Content-Type': 'application/json',
          Accept: 'application/json',
          'X-Requested-With': 'XMLHttpRequest',
          'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36',
        },
      })
        .then(({ data }) => {
          resolve({
            ...data.items[0].user,
            result: data.items[0].video_versions,
            thumbnail: data.items[0].image_versions2.candidates[0].url,
            caption: data.items[0].caption.text,
            views: data.items[0].view_count,
            duration: data.items[0].video_duration,
            like: data.items[0].like_count,
            comment: data.items[0].comment_count,
            music: data.items[0].clips_metadata?.music_info?.music_asset_info.title || '',
            artist: data.items[0].clips_metadata?.music_info?.music_asset_info.display_artist || '',
            url_mp3: data.items[0].clips_metadata?.music_info?.music_asset_info.progressive_download_url || '',
            url: url,
          })
        })
        .catch((err) => {
          reject(err)
        })
    })
  }
}
