// hikki-me/jamal-bot/src/lib/api/topup.ts
import axios from 'axios'
import FormData from 'form-data'
export class topUp {
  public async topUpFreeFire(props: free_fire): Promise<IResultPaymet> {
    return new Promise(async (resolve, reject) => {
      const form = new FormData()
      if (props.nominal == 5) {
        form.append('productId', '3')
        form.append('itemId', '353')
        form.append('catalogId', '376')
        form.append('paymentId', '2037')
        form.append('gameId', props.id)
        form.append('product_ref', 'REG')
        form.append('product_ref_denom', 'AE')
      } else if (props.nominal == 12) {
        form.append('productId', '3')
        form.append('itemId', '10')
        form.append('catalogId', '65')
        form.append('paymentId', '2039')
        form.append('gameId', props.id)
        form.append('product_ref', 'REG')
        form.append('product_ref_denom', 'AE')
      }
      await axios
        .post('https://api.duniagames.co.id/api/transaction/v1/top-up/inquiry/store', form, {
          headers: {
            'X-Requested-With': 'XMLHttpRequest',
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36',
            'Content-Type': 'multipart/form-data; boundary=' + form.getBoundary(),
            'Content-Length': form.getLengthSync().toString(),
          },
        })
        .then(({ data }) => {
          resolve(data)
        })
        .catch((err) => {
          reject(err)
        })
    })
  }
  public async getPaymentGateway(result: IResultPaymet, phoneNumber: string): Promise<payIsPay> {
    return new Promise(async (resolve, reject) => {
      if (!phoneNumber.startsWith('08')) return reject('Phone number must start with 08')
      const form = new FormData()
      form.append('inquiryId', result.data.inquiryId)
      form.append('phoneNumber', '081241195562')
      form.append('transactionId', result.data.transactionId)
      await axios
        .post('https://api.duniagames.co.id/api/transaction/v1/top-up/transaction/store', form, {
          headers: {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36',
            'Content-Type': 'multipart/form-data; boundary=' + form.getBoundary(),
            'Content-Length': form.getLengthSync().toString(),
          },
        })
        .then(({ data }) => {
          resolve(data)
        })
        .catch((err) => {
          reject(err)
        })
    })
  }
}
type inominalMl = 3 | 12 | 28 | 59 | 85 | 170 | 296 | 875
export type inominalFf = 5 | 12 | 50 | 70 | 140 | 355 | 720 | 1450
type free_fire = {
  id: string
  nominal: inominalFf
}
type mobile_legends = {
  id: string
  server: string
  nominal: inominalMl | number
}
type ItopUp = mobile_legends | free_fire
type IResultPayment = {
  status: string
  message: string
}
export type IResultPaymet = {
  data: {
    inquiryId: number
    gameId: string
    transactionId: number
    paymentId: number
    userNameGame: string
  }
}
type payIsPay = {
  deepLink: string
  qrCode: string
  checkoutUrl: string
  paymentMethod: 'QRIS'
  timerCount: number
}
